File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/store/CategoriesAdapter.java
package com.mlsdev.mlsdevstore.presentaion.store;

import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.mlsdev.mlsdevstore.R;
import com.mlsdev.mlsdevstore.data.model.category.CategoryTreeNode;
import com.mlsdev.mlsdevstore.databinding.ItemCategoryBinding;

import java.util.ArrayList;
import java.util.List;

public class CategoriesAdapter extends RecyclerView.Adapter<CategoriesAdapter.ViewHolder> {

    private List<CategoryTreeNode> categoryTreeNodes = new ArrayList<>();

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ItemCategoryBinding binding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()),
                R.layout.item_category,
                parent,
                false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.onBindView();
    }

    @Override
    public int getItemCount() {
        return categoryTreeNodes.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private ItemCategoryBinding binding;

        public ViewHolder(ItemCategoryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void onBindView() {
            CategoryTreeNode node = categoryTreeNodes.get(getAdapterPosition());

            if (binding.getViewModel() == null)
                binding.setViewModel(new CategoryItemViewModel());

            binding.getViewModel().setData(node);
        }
    }

    public void setData(List<CategoryTreeNode> categoryTreeNodes) {
        this.categoryTreeNodes.clear();
        this.categoryTreeNodes.addAll(categoryTreeNodes);
    }

}

File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/store/CategoriesAdapter.kt
package com.mlsdev.mlsdevstore.presentaion.store

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.mlsdev.mlsdevstore.R
import com.mlsdev.mlsdevstore.data.model.category.CategoryTreeNode
import com.mlsdev.mlsdevstore.databinding.ItemCategoryBinding
import java.util.*

class CategoriesAdapter(val onItemClick: (item: CategoryTreeNode) -> Unit) : RecyclerView.Adapter<CategoriesAdapter.ViewHolder>() {

    private val categoryTreeNodes = ArrayList<CategoryTreeNode>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = DataBindingUtil.inflate<ItemCategoryBinding>(
                LayoutInflater.from(parent.context),
                R.layout.item_category,
                parent,
                false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.onBindView()
    }

    override fun getItemCount(): Int {
        return categoryTreeNodes.size
    }

    inner class ViewHolder(private val binding: ItemCategoryBinding) : RecyclerView.ViewHolder(binding.root) {

        fun onBindView() {
            val node = categoryTreeNodes[adapterPosition]

            if (binding.viewModel == null)
                binding.viewModel = CategoryItemViewModel(onItemClick)

            binding.viewModel?.setData(node)
        }
    }

    fun setData(categoryTreeNodes: List<CategoryTreeNode>) {
        this.categoryTreeNodes.clear()
        this.categoryTreeNodes.addAll(categoryTreeNodes)
    }

}
